File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun maxPathSum(root: TreeNode?): Int {
        var res = Int.MIN_VALUE
        fun helper(root: TreeNode?): Int {
            if(root==null)
                return 0
            
            val l = helper(root.left)
            val r = helper(root.right)

            val lr_max = maxOf(l, r)
            val root_max= maxOf(root.`val`, root.`val` + lr_max)
            val all_max = maxOf(root_max, root.`val` +  + l+ r)

            res = maxOf(res, all_max)

            return root_max
        }
        helper(root)
        return res

    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        res = []
        def helper(root):
            if not root:
                res.append("None")
                return 
            
            res.append(str(root.val))
            helper(root.left)
            helper(root.right)
            
        helper(root)
        return ", ".join(res)


    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> TreeNode:
        data = data.split(", ")
        self.pos = 0

        def helper(root):
            if data[0] == "None":
                data.pop(0)
                return None

            root = TreeNode(int(data.pop(0)))
            root.left = helper(root)
            root.right = helper(root)

            return root
        return helper(None)